### PR TITLE
Replace legacy lock screen implementation

### LINKED ISSUE #6217: Replace PIN code screen implementation
#### What would you like to do?

Replace PF-LockScreen library with our own implementation.

#### Why would you like to do it?

While this library is really handy, the internal APIs used have been deprecated and modifying it can be difficult. It also lacks some features as key invalidation on new biometric method enrollment, which is needed by some of our customers.

#### How would you like to achieve it?

Re-implement this feature following our code guidelines, using the new APIs available and simplifying how it works internally.

### Have you considered any alternatives?

Modifying our PF-LockScreen fork was a possibility, but in the long term not depending on a library that's no longer maintained and uses deprecated APIs is probably the best way to go.

### Additional context

_No response_

### Are you willing to provide a PR?

Yes

### PR DESCRIPTION
**Pull request by jmartinesp**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Our PF-LockScreen fork is replaced by our own implementation of this feature. This should be easier to customise and maintain, it allows us to use the new biometric APIs from Android, removing the deprecated ones, and adds missing features like biometric key invalidation on new method enrolled.

In the beginning this was a separate module, but due to dependencies with some classes in the vector module it had to be included as part of this module. It's in its own `im.vector.app.features.pin.lockscreen` package so it can be split in the future. Sadly, the resources needed are mostly in the `library/ui-styles` module and are not as easy to split as they could be, but most of them are prefixed using `lockscreen_` to be easily distinguished.

## Motivation and context

Fixes #6217, depends on #6270 .

## Screenshots / GIFs

No UI changes, we're using the same resources and styles.

## Tests

- Have a logged in user.
- Go to Settings -> Security & Privacy -> Protect access.
- Enable PIN, set up a PIN code.
- Enable biometrics too if possible, confirm with a fingerprint (this can also be done on emulators from the settings dialog, but emulator with API 29 had some issues with fingerprints).
- To test this feature more easily, disable "Require PIN after 2 minutes" so the PIN code screen is displayed every time you return to the app.

Which things could be useful to test:
- Enabling / disabling biometrics.
- Failing the biometric prompt too many times, it should automatically disable biometric auth.
- Failing PIN code auth too many times should force a log out.
- Changing PIN.

Note: if the user updates the app and they already had both PIN code and biometric authentication enabled, _key invalidation will only start working after the first successful biometric auth_. A system/biometric key will be created and maintained. Also, key invalidation is only available since Android 6.

## Tested devices

- [x] Physical
- [x] Emulator
- OS version(s): 9, 10, 12 (both emulator and physical devices).

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: vector/src/androidTest/java/im/vector/app/features/pin/lockscreen/biometrics/BiometricHelperTests.kt [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    private val biometricManager = mockk<BiometricManager>(relaxed = true)
+    private val lockScreenKeyRepository = mockk<LockScreenKeyRepository>(relaxed = true)
+
+    @Before
+    fun setup() {
+        clearAllMocks()
+    }
+
+    @Test
+    fun canUseWeakBiometricAuthReturnsTrueIfIsFaceUnlockEnabledAndCanAuthenticate() {
```
Comment (ganfra) [inline vector/src/androidTest/java/im/vector/app/features/pin/lockscreen/biometrics/BiometricHelperTests.kt:63]: Would be more readable to use back-ticks like this:

```
fun `i am a test method`(){
  //
}
```

File: vector/src/main/java/im/vector/app/core/di/LockScreenModule.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            pinCodeLength = 4,
+            isWeakBiometricsEnabled = false,
+            isDeviceCredentialUnlockEnabled = false,
+            isStrongBiometricsEnabled = true,
+            needsNewCodeValidation = true,
+    )
+
+    @Provides
+    @Singleton
+    fun provideKeyHelper(@ApplicationContext context: Context) = LockScreenKeyRepository(context, "vector-lockscreen")
```
Comment (ganfra) [inline vector/src/main/java/im/vector/app/core/di/LockScreenModule.kt:55]: You should annotate your constructor of LockScreenKeyRepository directly with `@Inject` so you don't need to do it in module.
Also if it doesn't hold a state or init is not heavy it shouldn't be scoped

**Review by ganfra — CHANGES_REQUESTED** [review]
Nice work! 
Some remarks, mostly on the DI.
Also, could you use svg instead of xxxhdpi pngs?

File: vector/src/main/java/im/vector/app/core/di/LockScreenModule.kt (hunk 2 above)
Comment (jmartinesp) [reply vector/src/main/java/im/vector/app/core/di/LockScreenModule.kt:55]: About the repository being declared in the module, I did it like this to configure the `baseName` "vector-lockscreen" parameter. I could do it by using a default value for baseName or by declaring this value with some qualifier and then injecting it automatically, but it seems too complex for this.

As for state, it _could happen_ that 2 different badly used repositories created some key at the same time, that's why I used singleton here.

File: vector/src/main/java/im/vector/app/features/pin/lockscreen/ui/LockScreenAction.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+package im.vector.app.features.pin.lockscreen.ui
+
+import androidx.fragment.app.FragmentActivity
+import im.vector.app.core.platform.VectorViewModelAction
+
+sealed class LockScreenAction : VectorViewModelAction {
+    data class PinCodeEntered(val value: String) : LockScreenAction()
+    data class ShowBiometricPrompt(val callingActivity: FragmentActivity) : LockScreenAction()
```
[the remark this answers is not in the document]
Comment (jmartinesp) [reply vector/src/main/java/im/vector/app/features/pin/lockscreen/ui/LockScreenAction.kt:24]: We need to pass it to `BiometricHelper`,  so I don't think we can avoid doing this without changing the whole structure 😞.

File: vector/src/main/java/im/vector/app/features/pin/lockscreen/views/LockScreenCodeView.kt [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+    /**
+     * Number of digits entered.
+     */
+    val enteredDigits: Int get() = code.size
+
+    /**
+     * Callback called when the PIN code has been completely entered.
+     */
+    var onCodeCompleted: ((String) -> Unit)? = null
```
[the remark this answers is not in the document]
Comment (jmartinesp) [reply vector/src/main/java/im/vector/app/features/pin/lockscreen/views/LockScreenCodeView.kt:51]: Isn't it a bit redundant? It will get automatically translated to a lambda anyway. In any case, I'll change it if that's what we prefer.

File: vector/src/main/java/im/vector/app/core/di/LockScreenModule.kt (hunk 2 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/core/di/LockScreenModule.kt:55]: Ok for this class then

File: vector/src/main/java/im/vector/app/features/pin/lockscreen/views/LockScreenCodeView.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (ganfra) [reply vector/src/main/java/im/vector/app/features/pin/lockscreen/views/LockScreenCodeView.kt:51]: As you want, I don't have preference for this

File: vector/src/main/java/im/vector/app/features/pin/lockscreen/ui/LockScreenListener.kt [hunk 5]
```diff
@@ -19,45 +19,45 @@ package im.vector.app.features.pin.lockscreen.ui
 /**
  * Listener class to be notified of any event that could happen in the lock screen UI.
  */
-open class LockScreenListener {
+interface LockScreenListener {
     /**
      * In PIN creation mode, called when the first PIN code has been entered.
      */
-    open fun onFirstCodeEntered() {}
+    fun onFirstCodeEntered() {}
```
Comment (ganfra) [inline vector/src/main/java/im/vector/app/features/pin/lockscreen/ui/LockScreenListener.kt:26]: The default impl could just return Unit but it's a detail
`fun onFirstCodeEntered() = Unit`

File: vector/src/main/java/im/vector/app/features/pin/lockscreen/di/LockScreenModule.kt [hunk 6]
```diff
@@ -52,19 +54,17 @@ object LockScreenModule {
...
+    fun provideKeyRepository(@ApplicationContext context: Context) = LockScreenKeyRepository(context, "vector-lockscreen")
 
     @Provides
-    @Singleton
-    fun provideBiometricUtils(
+    fun provideBiometricHelper(
             @ApplicationContext context: Context,
             lockScreenKeyRepository: LockScreenKeyRepository,
             configurationProvider: LockScreenConfiguratorProvider,
     ) = BiometricHelper(context, lockScreenKeyRepository, configurationProvider, BiometricManager.from(context))
```
Comment (ganfra) [inline vector/src/main/java/im/vector/app/features/pin/lockscreen/di/LockScreenModule.kt:64]: Use Inject annotation for BiometricHelper and PinCodeHelper, no?

File: vector/src/main/java/im/vector/app/features/pin/lockscreen/ui/LockScreenFragment.kt [hunk 7]
```diff
@@ -164,7 +164,7 @@ class LockScreenFragment : VectorBaseFragment<FragmentLockScreenBinding>() {
 
     private fun setupCodeView(lockScreenCodeView: LockScreenCodeView, configuration: LockScreenConfiguration) = with(lockScreenCodeView) {
         codeLength = configuration.pinCodeLength
-        onCodeCompleted = { code ->
+        onCodeCompleted = LockScreenCodeView.CodeCompletedListener { code ->
```
Comment (ganfra) [inline vector/src/main/java/im/vector/app/features/pin/lockscreen/ui/LockScreenFragment.kt:167]: Use fonction reference?

**Review by ganfra — COMMENTED** [review]
Just small remaining questions

File: vector/src/main/java/im/vector/app/features/pin/lockscreen/ui/LockScreenFragment.kt (hunk 7 above)
Comment (jmartinesp) [reply vector/src/main/java/im/vector/app/features/pin/lockscreen/ui/LockScreenFragment.kt:167]: Sorry, I'm not following. How should I use a function reference here? By extracting its contents to private function first?

Comment (jmartinesp) [reply vector/src/main/java/im/vector/app/features/pin/lockscreen/ui/LockScreenFragment.kt:167]: Oh, ok. Honestly, I don't think its worth it, we'd just be moving the `viewModel.handle(LockScreenAction.PinCodeEntered(code))` code into a `handleCodeCompleted(code: String)` which probably doesn't add much value. If this was used in several places it could make sense.

**Comment by jmartinesp** [comment]
> Thanks for the changes, all good for me. Note that I didn't test it, do you prefer I do or are you confident enough?

I tested everything that came to mind, but honestly another round of testing for edge cases wouldn't hurt if you have the time.

File: library/ui-styles/src/main/res/values/styles_pin_code.xml [hunk 8]
```diff
@@ -22,13 +22,13 @@
     </style>
 
     <style name="PinCodeDeleteButtonStyle">
-        <item name="android:src">@drawable/delete_lockscreen_pf</item>
+        <item name="android:src">@drawable/lockscreen_delete</item>
```
[the remark this answers is not in the document]
Comment (jmartinesp) [reply library/ui-styles/src/main/res/values/styles_pin_code.xml:25]: Oops, true. @ganfra also mentioned this, but I forgot to re-check those. These drawables are taken from the old lib, but I'll try to replace them with svg if possible.
